### PR TITLE
Move pirated stuff from Nemo

### PR DESCRIPTION
**Pull request by lgoettgens**
Counterpart to https://github.com/Nemocas/Nemo.jl/pull/1531. This will most certainly be a breaking change.

### CODE REVIEW

File: src/NemoStuff.jl [hunk 1]
```diff
@@ -1,5 +1,7 @@
 export neg!
 
+sub!(z::T, x::T, y::T) where {T} = x - y
```
Comment (fingolfin) [inline src/NemoStuff.jl:9]: Should there be a similar method for `add!` and/or `mul!`?

Note that in `src/NCRings.jl` we already have this:
```julia
function zero!(z::NCRingElem)
   return zero(parent(z))
end

function add!(z::T, x::T, y::T) where T <: NCRingElem
   return x + y
end

function addeq!(z::T, y::T) where T <: NCRingElem
   return z + y
end

function sub!(z::T, x::T, y::T) where T <: NCRingElem
   return x - y
end

function mul!(z::T, x::T, y::T) where T <: NCRingElem
   return x*y
end

function addmul!(z::T, x::T, y::T, c::T) where T <: NCRingElem
   c = mul!(c, x, y)
   z = addeq!(z, c)
   return z
end

```
So it seems this already does the same thing, the new code here "only" drops the requirement `sub!` (why? probably to support `Int`, `Rational`, etc.?)

Comment (lgoettgens) [reply src/NemoStuff.jl:9]: I noticed that as well, but I would like to leave this PR simply as a copy-paste of stuff. There are a few similar things as well that should be cleaned up in a follow-up PR.

Comment (fingolfin) [reply src/NemoStuff.jl:9]: Sure.

But just to record my thoughts for the future: I suggest we drop this method here, and move those from `src/NCRings.jl` to `src/fundamental_interface.jl`, but making them generic, i.e. letting `T` be free.... Or indeed, letting all three arguments be "free":  `add!(x,a,b) = a+b` etc.

File: src/NemoStuff.jl [hunk 2]
```diff
@@ -137,7 +289,29 @@ function sub(M::MatElem{T}, r::AbstractUnitRange{<:Integer}, c::AbstractUnitRang
...
+end
+
+################################################################################
+#
+#  Kernel over different rings
+#
+################################################################################
+
+@doc raw"""
+kernel(a::MatrixElem{T}, R::Ring; side::Symbol = :right) -> n, MatElem{elem_type(R)}
```
Comment (fingolfin) [inline src/NemoStuff.jl:354]:
```suggestion
    kernel(a::MatrixElem{T}, R::Ring; side::Symbol = :right) -> n, MatElem{elem_type(R)}
```

**Comment by fingolfin** [comment]
Any idea why the `diagonal_matrix` doctest suddenly is failing?

File: src/NemoStuff.jl [hunk 3]
```diff
@@ -86,8 +117,129 @@ function is_symmetric(M::MatElem)
...
+end
+
+function is_zero_row(M::Matrix{T}, i::Int) where {T<:Integer}
+    for j = 1:Base.size(M, 2)
+        if !iszero(M[i, j])
+            return false
+        end
+    end
+    return true
+end
```
Comment (fingolfin) [inline src/NemoStuff.jl:202]:
```suggestion
end

# TODO: add `is_zero_column(M::Matrix{T}, i::Int) where {T<:Integer}`
```

(I know you want to keep this PR as simple as possible, but it's also a great opportunity to review this code...)

File: src/NemoStuff.jl [hunk 4]
```diff
@@ -86,8 +117,129 @@ function is_symmetric(M::MatElem)
...
+    end
+    return z
+end
+
+function identity_matrix(::Type{MatElem}, R::Ring, n::Int)
+    return identity_matrix(R, n)
+end
+
+function is_zero_row(M::Matrix{T}, i::Int) where {T<:Integer}
+    for j = 1:Base.size(M, 2)
```
Comment (fingolfin) [inline src/NemoStuff.jl:196]: I'd prefer this, but then we need `ncols(M::Matrix) = Base.size(M, 2)` (which of course we already do have in `Nemo.jl/src/HeckeMoreStuff.jl:1711`...)
```suggestion
    for j = 1:ncols(M)
```

**Comment by lgoettgens** [comment]
> Any idea why the `diagonal_matrix` doctest suddenly is failing?

Flint's QQElems print differently than AbstractAlgebra's QQElems. And since flint is not available here, `QQ` refers to a different implementation of the rationals.

File: src/NemoStuff.jl (hunk 4 above)
Comment (lgoettgens) [reply src/NemoStuff.jl:196]: `ncols` is part of what is being moved, so no problem in applying this change.

File: src/NemoStuff.jl (hunk 3 above)
Comment (lgoettgens) [reply src/NemoStuff.jl:202]:
```suggestion
end

# TODO: add `is_zero_column(M::Matrix{T}, i::Int) where {T<:Integer}` and specialized functions in Nemo
```

**Comment by lgoettgens** [comment]
GitHub seems to have some issues with tar files in CI. I will wait some time and then rerun all jobs.

**Comment by fingolfin** [comment]
> > Any idea why the `diagonal_matrix` doctest suddenly is failing?
> 
> Flint's QQElems print differently than AbstractAlgebra's QQElems. And since flint is not available here, `QQ` refers to a different implementation of the rationals.

Yeah, but the tests were already here and not failing, which confused me. But I guess your PR #1413 answers that... ;-)

**Comment by lgoettgens** [comment]
> I am wondering now, will this technically be a breaking change? We "only" add methods; sure, some are overwritten in Nemo (and perhaps elsewhere), but by themselves they should not break anything, right?

I broke at least `vcat(Vector{MatElem})` for everyone using AbstractAlgebra and not Nemo. So yes, I would consider this breaking.

File: src/NemoStuff.jl [hunk 5]
```diff
@@ -0,0 +1,804 @@
...
+neg!(z::Rational{Int}, x::Rational{Int}) = -x
+
+add!(z::Rational{Int}, x::Rational{Int}, y::Int) = x + y
+
+mul!(z::Rational{Int}, x::Rational{Int}, y::Int) = x * y
+
+is_negative(x::Rational) = x.num < 0
+
+is_negative(n::Integer) = cmp(n, 0) < 0
+is_positive(n::Integer) = cmp(n, 0) > 0
```
Comment (fingolfin) [inline src/NemoStuff.jl:22]: For a follow-up PR:
```suggestion
is_negative(n::T) where T<:Real = n < zero(T)
is_positive(n::T) where T<:Real = n > zero(T)
```
Of course that ignores some icky questions, like: should `is_negative(-0.0)` be `true` (matching `signbit(-0.0) == true`) or `false` (matching `-0.0 == 0.0` resp. `iszero(-0.0)` both holding)? But I think for our purposes the above makes most sense.

File: src/NemoStuff.jl [hunk 6]
```diff
@@ -0,0 +1,804 @@
...
+function is_symmetric(M::MatElem)
+    for i in 1:nrows(M)
+        for j in i:ncols(M)
+            if M[i, j] != M[j, i]
+                return false
+            end
+        end
+    end
+    return true
+end
```
Comment (fingolfin) [inline src/NemoStuff.jl:148]: We already have the following in `src/Matrix.jl:1302` -- these should be merged (probably by just dropping the restriction on `T` on the existing method)
```
function is_symmetric(a::MatrixElem{T}) where T <: NCRingElement
    if !is_square(a)
        return false
    end
    for row in 2:nrows(a)
        for col in 1:(row - 1)
            if a[row, col] != a[col, row]
                return false
            end
        end
    end
    return true
end
```

File: src/NemoStuff.jl [hunk 7]
```diff
@@ -0,0 +1,804 @@
...
+
+################################################################################
+#
+#  Concatenation of matrices
+#
+################################################################################
+
+@doc raw"""
+vcat(A::Vector{Mat}) -> Mat
```
Comment (fingolfin) [inline src/NemoStuff.jl:373]:
```suggestion
    vcat(A::Vector{Mat}) -> Mat
```

File: src/NemoStuff.jl [hunk 8]
```diff
@@ -0,0 +1,804 @@
...
+
+function canonical_unit(a::SeriesElem)
+    iszero(a) && return one(parent(a))
+    v = valuation(a)
+    v == 0 && return a
+    v > 0 && return shift_right(a, v)
+    return shift_left(a, -v)
+end
+
+# should be Nemo/AA
```
Comment (fingolfin) [inline src/NemoStuff.jl:595]:
```suggestion
```

File: src/NemoStuff.jl [hunk 9]
```diff
@@ -0,0 +1,804 @@
...
+end
+
+ngens(R::MPolyRing) = nvars(R)
+
+Random.gentype(::Type{T}) where {T<:FinField} = elem_type(T)
+
+import LinearAlgebra
+LinearAlgebra.dot(a::RingElem, b::RingElem) = a * b
+
+function is_upper_triangular(M::MatElem)
```
Comment (fingolfin) [inline src/NemoStuff.jl:746]: To be merged with `is_upper_triangular` method in `src/Matrix.jl:3359`

File: src/NemoStuff.jl [hunk 10]
```diff
@@ -0,0 +1,804 @@
...
+
+transpose!(A::MatrixElem) = transpose(A)
+
+function Base.div(f::PolyRingElem, g::PolyRingElem)
+    q, r = divrem(f, g)
+    return q
+end
+
+function Base.rem(f::PolyRingElem, g::PolyRingElem)
+    return mod(f, g)
```
Comment (fingolfin) [inline src/NemoStuff.jl:766]: Huh, I wonder why not
```suggestion
    q, r = divrem(f, g)
    return r
```

**Review by fingolfin — COMMENTED** [review]
What's the plan, when will this be non-draft?

I propose that we merge & release soon, together with PR #1410, as 0.32, and then quickly proceed with updating Nemo, Hecke, Oscar, ...

(Unless e.g. @thofma or @fieker have objections?)

File: src/NemoStuff.jl [hunk 11]
```diff
@@ -536,10 +536,6 @@ function change_base_ring(p::MPolyRingElem{T}, g, new_polynomial_ring) where {T<
     return finish(M)::elem_type(new_polynomial_ring)
 end
 
-function leading_coefficient(f::Generic.MPoly)
-    return f.coeffs[1]
-end
-
```
Comment (lgoettgens) [inline src/NemoStuff.jl]: This fails for the zero polynomial. AA already contains a better implementation.

File: src/NemoStuff.jl [hunk 12]
```diff
@@ -560,18 +556,6 @@ function polynomial_ring(R::Ring; cached::Bool=false)
...
-    l = length(f)
-    if l == 0
-        return f
-    end
-    A = f.exps
-    r, c = size(A)
-    e = A[1:r, 1:1]
-    return R([one(base_ring(R))], e)
-end
-
```
Comment (lgoettgens) [inline src/NemoStuff.jl]: same here, fails for the zero poly

File: test/generic/FunctionField-test.jl [hunk 13]
```diff
@@ -98,14 +98,14 @@ end
...
+      @test_broken false # test_rand(S, 1:10, -10:10) # TODO: make this work again after #1412
    end
-   
+
    for f in P2
       S, y = FunctionField(f, "y")
 
       # TODO: test more than just the result type
-      test_rand(S, 1:10)
+      @test_broken false # test_rand(S, 1:10) # TODO: make this work again after #1412
```
Comment (lgoettgens) [inline test/generic/FunctionField-test.jl:108]: I have no idea how and why this broke. If you have an idea, please let me know.

**Comment by lgoettgens** [comment]
AbstractAlgebra tests now succeed locally on julia-1.9. Let's see if that's the case here as well, and the OscarCI-matching jobs as well.

**Comment by thofma** [comment]
I am happy with a quick 0.32 release.

File: test/generic/FunctionField-test.jl (hunk 13 above)
Comment (fingolfin) [reply test/generic/FunctionField-test.jl:108]: Argh I wrote a bunch of comments here and my browser died *sigh*. Let's see if I can cobble it back together:

As far as I can tell, there is already a problem here in `master`... but somehow there I can only trigger it when I handle the test file "interactively". That is, if I enter the following, I get an error also on `master` (I am assuming this code is executed from inside the `AbstractAlgebra.jl` directory:
```
using AbstractAlgebra, Test
include("test/rand.jl")
include("test/generic/FunctionField-test.jl")
```
this gives me:
```
...
Test Summary:                  | Pass  Total  Time
Generic.FunctionField.printing |    9      9  0.0s
Generic.FunctionField.rand: Error During Test at /Users/mhorn/Projekte/OSCAR/AbstractAlgebra.jl/test/generic/FunctionField-test.jl:96
  Got exception outside of a @test
  return type AbstractAlgebra.Generic.FunctionFieldElem{Rational{BigInt}} does not match inferred return type Any
  Stacktrace:
    [1] error(s::String)
      @ Base ./error.jl:35
    [2] test_rand(::Union{Nothing, Function}, ::Any, ::Any, ::Vararg{Any}; type::Any)
      @ Main ~/Projekte/OSCAR/AbstractAlgebra.jl/test/rand.jl:32
    [3] test_rand
      @ ~/Projekte/OSCAR/AbstractAlgebra.jl/test/rand.jl:18 [inlined]
    [4] #test_rand#9
      @ ~/Projekte/OSCAR/AbstractAlgebra.jl/test/rand.jl:15 [inlined]
    [5] test_rand(::Any, ::Any, ::Any)
      @ Main ~/Projekte/OSCAR/AbstractAlgebra.jl/test/rand.jl:15
    [6] macro expansion
      @ ~/Projekte/OSCAR/AbstractAlgebra.jl/test/generic/FunctionField-test.jl:101 [inlined]
    [7] macro expansion
      @ ~/.julia/juliaup/julia-1.9.3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.9/Test/src/Test.jl:1498 [inlined]
    [8] top-level scope
      @ ~/Projekte/OSCAR/AbstractAlgebra.jl/test/generic/FunctionField-test.jl:97
    [9] include(fname::String)
      @ Base.MainInclude ./client.jl:478
   [10] top-level scope
      @ REPL[17]:1
   [11] eval
      @ ./boot.jl:370 [inlined]
   [12] eval_user_input(ast::Any, backend::REPL.REPLBackend, mod::Module)
      @ REPL ~/.julia/juliaup/julia-1.9.3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.9/REPL/src/REPL.jl:153
   [13] repl_backend_loop(backend::REPL.REPLBackend, get_module::Function)
      @ REPL ~/.julia/juliaup/julia-1.9.3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.9/REPL/src/REPL.jl:249
   [14] start_repl_backend(backend::REPL.REPLBackend, consumer::Any; get_module::Function)
      @ REPL ~/.julia/juliaup/julia-1.9.3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.9/REPL/src/REPL.jl:234
   [15] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool, backend::Any)
      @ REPL ~/.julia/juliaup/julia-1.9.3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.9/REPL/src/REPL.jl:379
   [16] run_repl(repl::REPL.AbstractREPL, consumer::Any)
      @ REPL ~/.julia/juliaup/julia-1.9.3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.9/REPL/src/REPL.jl:365
   [17] (::Base.var"#1017#1019"{Bool, Bool, Bool})(REPL::Module)
      @ Base ./client.jl:421
   [18] #invokelatest#2
      @ ./essentials.jl:819 [inlined]
   [19] invokelatest
      @ ./essentials.jl:816 [inlined]
   [20] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
      @ Base ./client.jl:405
   [21] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:322
   [22] _start()
      @ Base ./client.jl:522
Test Summary:              | Error  Total  Time
Generic.FunctionField.rand |     1      1  0.1s
ERROR: LoadError: Some tests did not pass: 0 passed, 0 failed, 1 errored, 0 broken.
in expression starting at /Users/mhorn/Projekte/OSCAR/AbstractAlgebra.jl/test/generic/FunctionField-test.jl:96
```


And this I tracked down further to
```
using AbstractAlgebra, Test
using AbstractAlgebra.RandomExtensions: RandomExtensions, make
R = polynomial_ring(polynomial_ring(QQ,:x1)[1], :y)[1]
@inferred make(R, 0:0, 1:10, -10:10)
```
But `@code_warntype make(R, 0:0, 1:10, -10:10)` reveals nothing, no code ambiguity?  It also infers the correct result type. Huh, weird.

Comment (fingolfin) [reply test/generic/FunctionField-test.jl:108]: PR #1420 fixes this, so in theory after merging this we can re-run the tests here and hopefully they'll pass

Comment (lgoettgens) [reply test/generic/FunctionField-test.jl]: Now that #1420 is merged, I reverted the commenting out. Let's hope it succeeds now.

**Review by fingolfin — APPROVED** [review]
Fine by me, thanks!

The NegInf stuff could be moved in a follow-up PR, if you are willing. Then we can merge #1410 and #1419 and then quickly make a breaking release, then update Nemo etc....

**Comment by lgoettgens** [comment]
Fine with me, let's get this going before there arise conflicts. Infinity stuff goes in a new PR

**Comment by lgoettgens** [comment]
@fingolfin Auto-merge failed as the OscarCI-matching job here has a different name than the `required` one.

**Pull request merged**